### PR TITLE
The Messenger: content update

### PR DESCRIPTION
**Pull request by alwaysintreble**
## What is this fixing or adding?
Adds portal shuffle, limited movement, and early meditation options. Adds Notes and Phobekins location name groups for exclude/priority stuff. Adds a launcher component which can auto install and launch the mod. Also removes the option for seals to be shuffled; they're always shuffled now, and oob logic - since i lack the knowledge to write this in the new region system.

Currently commented out in this branch but tested and functional; also adds transition shuffle - this needs the generic ER branch - and lets the launcher component pass the uri to the client allowing it to auto connect - this needs #2779

## How was this tested?
lots of private testing as well as any builds and runs distributed to the player base. Also added unit tests for the portal logic. ER still probably needs some unit tests but as that's not currently functional with this PR can be added at a later date.

## If this makes graphical changes, please attach screenshots.

messenger launcher

### CODE REVIEW

**Review by benny-dreamly — APPROVED** [review]

File: worlds/messenger/client_setup.py
Comment (nbrochu) [inline worlds/messenger/client_setup.py:163]: You never set back the working directory to Archipelago's root directory, which means that everything from this point forward uses the game's directory as the base path. This breaks everything else in the launcher.

Examples:

Launching another game client
image

Logs ending up in the game directory
image

**Review by nbrochu — COMMENTED** [review]
* Combed through all the Python files.
* Looked at the documentation. I would have pointed out the same stuff as Porygone there.
* Generated a seed using most of the new options (`early_meditation` and `shuffle_portals`) and another game
* PENDING! Played through that seed to completion without a problem

I found a really important issue in `client_setup.py` that absolutely needs fixing before considering merging this. Outside of that, great quality of life to do all this from the launcher!

I also really don't like that ER is here but commented out and depending on another PR that may or not get in before this one. Creates an ambiguous review context.

File: worlds/messenger/__init__.py
Comment (alwaysintreble) [reply worlds/messenger/__init__.py:257]: shuffled portals generally open the game up more than the vanilla layout does. the validation step will always pass if transitions are shuffled (though that's disabled in this current PR), and if transitions aren't shuffled it always succeeded on the first try in my testing. it's still *possible* that since it does the selection completely randomly that it might come up with an invalid selection, but it's still very unlikely. the failsafe is mostly just here in case someone plandoes 5-6 of the available 6 portals and none of them give any sphere 1 access.

File: worlds/messenger/rules.py
Comment (alwaysintreble) [reply worlds/messenger/rules.py:488]: It's going to return at some point. Someone who plays oob has already written out the logic for it, I just need to write it all into the logic systems here still. Challenging logic will probably also return at some point, but we haven't come up with enough changes to give it a distinct identity from hard logic.

File: worlds/messenger/subclasses.py
Comment (alwaysintreble) [reply worlds/messenger/subclasses.py:29]: it's part of generic ER placement logic.

File: worlds/messenger/__init__.py
Comment (alwaysintreble) [reply worlds/messenger/__init__.py:141]: yeah i was mixed on doing this. probably better to just do it more "correctly"

File: worlds/messenger/__init__.py
Comment (alwaysintreble) [reply worlds/messenger/__init__.py:322]: the main reason i didn't is because the features were kind of developed in tandem so i don't have a clean git history to just move them with. moving them out of here completely wouldn't be too difficult to do, though a bit of a pain to attempt to keep it clean. would prefer to leave it as is, but will remove it if a core maintainer says to.

File: worlds/messenger/client_setup.py
Comment (nbrochu) [inline worlds/messenger/client_setup.py:129]: I'm getting a prompt for a new mod update even though I already have the latest one. A SemVer string is not numeric (at least in Python 3.11) so `available_mod_update` always returns True.

image

File: worlds/messenger/__init__.py
Comment (alwaysintreble) [reply worlds/messenger/__init__.py:168]: i don't like creating variables for things i'm only going to use a single time if i can reasonably go without. this is a case where i agree the comprehension shouldn't have been here, though i will be keeping it (because like you said i find it perfectly legible).

File: worlds/messenger/__init__.py
Comment (alwaysintreble) [reply worlds/messenger/__init__.py:187]: this is a tiny bit slower, but the real reason i didn't do this is because i thought i'd need to move all the other items further to the right making the comprehension unreadable. IDE seems to want to move them further to the left though so using it.

File: worlds/messenger/__init__.py
Comment (alwaysintreble) [reply worlds/messenger/__init__.py:229]: originally this did just call `get_filler()` and `get_filler_item_name` had a larger arbitrary value because i usually need to create somewhere around 50-80 items here. berserker commented that it was wasteful to make my cache that large though so i opted to just not use the cache here so it doesn't have to constantly rebuild. the main reason i make the cache at all is because even though that only really happens if core calls it, core is almost never going to call your `get_filler()` only a single time and i'd rather not do the random calculation every single time it gets called.

File: worlds/messenger/options.py
Comment (nbrochu) [inline worlds/messenger/options.py:52]: I have my available portals set to 6 (the default) and have portal shuffle: anywhere set. I start the game and only get 3. The code in `generate_early` around available portals seems fine, so it's either a game mod issue, or the docstring for that option is not clear enough if it's an expected restriction with my options

image

image

Comment (alwaysintreble) [reply worlds/messenger/options.py:52]: the remaining portals open after you return to the HQ. this is intentional because I had some players enter portals without talking to the prophet, and I can't open all 6 portals in the cutscene.

File: worlds/messenger/client_setup.py
Comment (alwaysintreble) [reply worlds/messenger/client_setup.py:163]: whoops good catch.

**Review by nbrochu — APPROVED** [review]
My feedback was addressed, and the issues were fixed. Completed my test seed successfully too

File: worlds/messenger/__init__.py
Comment (black-sliver) [inline worlds/messenger/__init__.py:45]: > Adds Notes and Phobekins location name groups for exclude/priority stuff

I'm confused. The groups where already here?

File: worlds/messenger/__init__.py
Comment (black-sliver) [inline worlds/messenger/__init__.py:253]:
```python
else:
    ...
# else:
    ...
```
seems odd even for "this may get back in". I think if you want to not forget about the code path here, I'd keep the `elif` and then throw an "Unsupported logic" exception in the `else`.

File: worlds/messenger/client_setup.py
Comment (black-sliver) [inline worlds/messenger/client_setup.py:23]: As below, this should also work on mac. There are a few more instance of that here.

File: worlds/messenger/__init__.py
Comment (alwaysintreble) [reply worlds/messenger/__init__.py:257]: it's random fill and there's only 6 portals so it's really fast. even then it should only attempt more than once if the player uses plando, which is what the failsafe is for; if the player uses plando and creates a combo it can't solve.

File: worlds/messenger/client_setup.py
Comment (alwaysintreble) [reply worlds/messenger/client_setup.py:157]: i had no idea if it would also work on mac which is why i was explicit about not supporting it, but if it should work the same way i'll add that in. for linux https://discord.com/channels/731205301247803413/1090822939697500243/1215326568348389426

File: worlds/messenger/__init__.py
Comment (alwaysintreble) [reply worlds/messenger/__init__.py:168]: yeah i put this here because i didn't like that it said "generating output for 20 players" and then didn't actually output any files. thought i had removed it though. it's moved to a classmethod now.

File: worlds/messenger/__init__.py
Comment (black-sliver) [inline worlds/messenger/__init__.py:354]: You forgot to add a comment explaining what this is. it's super confusing if I didn't know already.

File: worlds/messenger/client_setup.py
Comment (black-sliver) [reply worlds/messenger/client_setup.py]: It's hard to say how long stuff is gonna work on mac, with mono being deprecated, and also I have no idea if courier actually works, but both [mono](https://formulae.brew.sh/formula/mono) as well as steam is available on mac.

**Review by black-sliver — APPROVED** [review]
I think this is fine now.

Thanks!

**Pull request merged**
